pulsar-io/azure-data-explorer/src/test/java/org/apache/pulsar/io/azuredataexplorer/ADXSinkE2ETest.java

```diff
@@ -0,0 +1,183 @@
+        String appId = System.getenv("kustoAadAppId");
+        String appkey = System.getenv("kustoAadAppSecret");
+
+        ConnectionStringBuilder engineKcsb =
+                ConnectionStringBuilder.createWithAadApplicationCredentials(ADXSinkUtils.getQueryEndpoint(cluster),
+                        appId, appkey, authorityId);
+        kustoAdminClient = ClientFactory.createClient(engineKcsb);
+
+        kustoAdminClient.execute(database, generateAlterIngestionBatchingPolicyCommand(database,
+                "{\"MaximumBatchingTimeSpan\":\"00:00:10\", \"MaximumNumberOfItems\": 500, \"MaximumRawDataSizeMB\": 1024}"));
+
+        String createTableCommand = ".create table " + table +
+                " ( key:string , value:string, eventTime:datetime , producerName:string , sequenceId:long ,properties:dynamic )";
+        kustoAdminClient.execute(database, createTableCommand);
+    }
+
+    private String generateAlterIngestionBatchingPolicyCommand(String entityName, String targetBatchingPolicy) {
+        return ".alter database " + entityName + " policy ingestionbatching ```" + targetBatchingPolicy + "```";
+    }
+
+    @AfterMethod(alwaysRun = true)
+    public void tearDown() {
+        try {
+            kustoAdminClient.execute(".drop table " + table + " ifexists");
+        } catch (Exception ignore) {
+        }
+    }
+    @Test
+    public void TestOpenAndWriteSink() throws Exception {
+
+
+        Map<String, Object> configs = new HashMap<>();
+        configs.put("clusterUrl", System.getenv("kustoCluster"));
+        configs.put("database", System.getenv("kustoDatabase"));
+        configs.put("table", table);
+        configs.put("batchTimeMs", 1000);
+        configs.put("flushImmediately", true);
+        configs.put("appId", System.getenv("kustoAadAppId"));
+        configs.put("appKey", System.getenv("kustoAadAppSecret"));
+        configs.put("tenantId", System.getenv("kustoAadAuthorityID"));
+        configs.put("maxRetryAttempts", 3);
+        configs.put("retryBackOffTime", 100);
+        /*configs.put("serviceUrl","pulsar://localhost:3000");
+        configs.put("dlqTopic","something-DLQ");*/
+
+        ADXSink sink = new ADXSink();
+        sink.open(configs, null);
+        int writeCount = 50;
+
+        for (int i = 0; i < writeCount; i++) {
+            Record<byte[]> record = build("key_" + i, "test data from ADX Pulsar Sink_" + i);
+            sink.write(record);
+        }
+        Thread.sleep(40000);
+        KustoOperationResult result = kustoAdminClient.execute(database, table + " | count");
+        KustoResultSetTable mainTableResult = result.getPrimaryResults();
+        mainTableResult.next();
+        int actualRowsCount = mainTableResult.getInt(0);
+        Assert.assertEquals(actualRowsCount, writeCount);
+
+        sink.close();
+    }
+
+    @Test
+    public void TestOpenAndWriteSinkWithTimeouts() throws Exception {
+
+
+        Map<String, Object> configs = new HashMap<>();
```

Comment:
Map can be set once and then re-used ? so that you need not do System.getEnv repeatedly ?